### ISSUE TITLE
Are the TwisterCrypto function working? If yes, how to use them?

### ISSUE THREAD

**Issue by Tschaul** [miguelfreitas/twister-html#239]
Hi,

For the twister library i'm writing at http://github.com/Tschaul/LibTwister that will at some point hopefully also be part of twister-html i implemented most of the passive functionality (things that can also be done as "guest"). For the active things (posting, updating profiles etc.) i would very much first implement the client-side-crypto mode and later add server-side-crypto functionality because client side crypto will be much more constraining for the overall design. That is why i ask to what extend the functions in TwisterCrypto already are usable and if so how exactly to use them e.g. to verify a post signature.

Greetings

**Comment by Tschaul** [comment]
I just found the [twister-crypto-test.html](https://github.com/miguelfreitas/twister-html/blob/master/twister-crypto-test.html). I will work from there.

**Comment by miguelfreitas** [comment]
Exactly! I believe twister-crypto-test.html should provide a minimum set of examples to get going.

There are minor details besides the crypto itself, as the structs must be uniquely "bencoded" (that is, bencoded with sorted dictionaries) before computing hashes and signatures. The examples are passing the testvector tests, so afaik everything we need is working fine.

Also notice there is still a missing command in twister-core to be able to submit "raw" (hex and already signed) posts to the torrent. The equivalent functions for sending "raw" DHT are already implemented and tested.

**Comment by Tschaul** [comment]
Whats also missing is a way to access the raw direct messages.

**Comment by Tschaul** [comment]
Verifying user posts works flawlessly. But I can get it to correctly verify the "outer" signature of the dhtpayload. I added a (not) working example in a branch of my twister-html https://github.com/Tschaul/twister-html/blob/more-crypto-testing/twister-crypto-test.html#L113 What am i doing wrong?

**Comment by Tschaul** [comment]
Retwisting posts also can not be verified, but the encapsulated retwisted post can... added examples to the same file: https://github.com/Tschaul/twister-html/blob/more-crypto-testing/twister-crypto-test.html#L113

**Comment by miguelfreitas** [comment]
About the access to raw direct messages: indeed, you are right. My thought was to obtain those using the same "getposts" rpc, the underling code is about the same, we just need to play a little bit with flags. Maybe adding another parameter will be required.

I have yet to see why your "outer" verification is failing. But I must sleep now and i have a long work day tomorrow... 

Just a raw shot: have you converted inner sig_userpost back from hex to binary before bencoding?

**Comment by Tschaul** [comment]
no hurry! I will try some more. Got lots of other stuff to implement too.

**Comment by miguelfreitas** [comment]
obtaining raw direct messages is now possible: just use getposts with allowed_flags=required_flags=2.

**Comment by Tschaul** [comment]
Thanks! Works, but why allowed and required? and why two fields and why 2 as value?

And yes you were right with the encoding of the inner signatures. It does work now. 

Another note: the signature verification takes quite a lot of cpu power. To verify 24 hours of posts from my timeline it takes about 1min and my machine is not slow at all. On a mobile phone this will be multiple minutes. So the signature verification must happen in the background without blocking the reponsiveness. That is how i implemented it: https://github.com/Tschaul/twister-lib-js/blob/master/src/TwisterPubKey.js#L88. I will, however, add a option for instant signature verification as well. Then same will probably be true for decrypting messages.

**Comment by miguelfreitas** [comment]
@Tschaul i'm keeping compatibility with the previous existing flag, which was used to filter posts and had the "allowed" semantic:
```
 (m_picker->post_flags(i) & allowed_flags) == m_picker->post_flags(i) &&
```
that is, if post has any flag which is not in "allowed" set, it is not returned.
then i've added another check with "required" meaning:
```
(m_picker->post_flags(i) & required_flags) == required_flags
```
so if post does not contains all of "required" set, it is not returned.

we only have normal (no flag), RT = 1 and DM = 2. But we may add new types, i don't know (group chat invite, for example)...

**Comment by miguelfreitas** [comment]
About the signature verification in javascript: I have noticed the slowness as well. For a while I've run a modified twister-html checking all posts in my timeline (to identify any bugs or verification failures). Updating posts on screen got noticeably slower :-(

I also thought that solution would be to make this asychronous signature checking, so i'm happy that you implemented it in background. nice!

**Comment by miguelfreitas** [comment]
"newpostraw" RPC implemented in twister-core. Tested with a simple post.
The downside of using newpostraw is that we need to redo all the work that newpostmsg usually does  inside twister-core for us, that is:
1) formatting and signing post (including all relevant fields, like msg, rt, k, time, height, lastk etc)
2) signing and posting to DHT as well ("postN" and "status" resources)
3) updating "repliesN" DHT resources of other users.
4) checking for mentions/hashtags and updating the respective DHT resources.

**Comment by Tschaul** [comment]
End of next week i start my new job. From then on things will go slower again from my side. I hope to implement the server-side-wallet part fully and have a basic level of documentation. Right now i'm tinkering with my react based client which helps a lot to find flaws and bugs. The wallet part is, however, fully modularized, so its easy to add in other classes which have the same interface, whether that is a full fledged client-side wallet or an interface to an extension-based wallet. For twister-html it would probably make most sense to wait until extenion-based or client-side crypto is fully implemented and tested before bringing it in. No need to change things without adding extra functionality. There is no need to hurry adding all the "raw" JSON-RPC. There is enough waiting to be implemented already.
